txs/conservative_state.go

```diff
@@ -146,67 +155,98 @@ func (cs *ConservativeState) RevertState(revertTo types.LayerID) (types.Hash32,
 }
 
 // ApplyLayer applies the transactions specified by the ids to the state.
-func (cs *ConservativeState) ApplyLayer(toApply *types.Block) ([]*types.Transaction, error) {
+func (cs *ConservativeState) ApplyLayer(toApply *types.Block) ([]types.TransactionID, error) {
 	logger := cs.logger.WithFields(toApply.LayerIndex, toApply.ID())
 	logger.Info("applying layer to conservative state")
 
 	if err := cs.cache.CheckApplyOrder(toApply.LayerIndex); err != nil {
 		return nil, err
 	}
 
-	txs, err := cs.getTXsToApply(toApply)
+	txs, raw, err := cs.getTXsToApply(toApply)
 	if err != nil {
 		return nil, err
 	}
 
-	failedTxs, err := cs.vmState.ApplyLayer(toApply.LayerIndex, txs, toApply.Rewards)
+	skipped, err := cs.vmState.Apply(toApply.LayerIndex, raw, toApply.Rewards)
 	if err != nil {
 		logger.With().Error("failed to apply layer txs",
 			toApply.LayerIndex,
-			log.Int("num_failed_txs", len(failedTxs)),
+			log.Int("num_failed_txs", len(skipped)),
 			log.Err(err))
 		// TODO: We want to panic here once we have a way to "remember" that we didn't apply these txs
 		//  e.g. persist the last layer transactions were applied from and use that instead of `oldVerified`
 		return nil, fmt.Errorf("apply layer: %w", err)
 	}
 
 	finalList := txs
-	if len(failedTxs) > 0 {
+	if len(skipped) > 0 {
 		finalList = make([]*types.Transaction, 0, len(txs))
 		failed := make(map[types.TransactionID]struct{})
-		for _, tx := range failedTxs {
-			failed[tx.ID()] = struct{}{}
+		for _, id := range skipped {
+			failed[id] = struct{}{}
 		}
 		for _, tx := range txs {
-			if _, ok := failed[tx.ID()]; !ok {
+			if _, ok := failed[tx.ID]; !ok {
 				finalList = append(finalList, tx)
 			}
 		}
 	}
 
 	logger.With().Info("applying layer to cache",
-		log.Int("num_txs_failed", len(failedTxs)),
+		log.Int("num_txs_failed", len(skipped)),
 		log.Int("num_txs_final", len(finalList)))
 	if _, errs := cs.cache.ApplyLayer(toApply.LayerIndex, toApply.ID(), finalList); len(errs) > 0 {
 		return nil, errs[0]
 	}
-	return failedTxs, nil
+	return skipped, nil
 }
 
-func (cs *ConservativeState) getTXsToApply(toApply *types.Block) ([]*types.Transaction, error) {
+func (cs *ConservativeState) getTXsToApply(toApply *types.Block) ([]*types.Transaction, []types.RawTx, error) {
+	// TODO this allocates an additional map for all tx ids in the block
+	// it would make much more sense to load transactions in the loop below
+	// and terminate with error if transaction is missing
```

Comment:
i don't understand this comment. what map are you referring to?

Reply:
i will remove. i don't recall, but there were unnecessary creation of the maps and arrays somewhere